MH-12502, Do Not Leave Files In Workspace

lkiesow opened the pull request:
- MH-12502, Convert Workspace Read To InputStream

    This patch makes the `workspace.read()` method return an InputStream,
    ensuring that the files content which may be read directly from the
    working file repository (if possible) is not modified.

    In case the files cannot be read directly (e.g. non-local files), the
    files are temporarily copied into the workspace, but will automatically
    be removed once the returned InputStream is closed to not leave any
    files lying around in the workspace.

- MH-12502, Ensure search service does not leave MPEG7 catalogs in workspace

    This patch ensures that MPEG7 catalogs, temporary loaded into the
    workspace for parsing will be removed afterward.

- MH-12502, Ensure OAI-PMH does not leave media package elements in WS/WFR

    This patch ensures that media package elements temporary loaded into the
    workspace for parsing will be removed afterward.

- MH-12502, Ensure Asset Manager Cleans Up Workspace

    This patch ensures that at several places the asset manager will clean
    up files it loads into the workspace.

- MH-12502, Add Options To Get Unique Files From Workspace
    
    This patch adds an optional `uniqueFilename` parameter to
    `workspace.get()`, allowing for the requested file to be stored in a
    unique location which makes it safe to modify the files content as well
    as deleting the file when it is not used anymore.

lkiesow committed:
MH-12502, Ensure OAI-PMH does not leave media package elements in WS/WFR

This patch ensure that media package elements temporary loaded into the
workspace for parsing will be removed afterwards.

lkiesow committed:
MH-12502, Ensure search service does not leave MPEG7 catalogs in workspace

This patch ensure that MPEG7 catalogs, temporary loaded into the
workspace for parsing will be removed afterwards.

lkiesow committed:
MH-12502, Ensure Asset Manager Cleans Up Workspace

This patch ensures that at several places the asset manager will clean
up files it loads into the workspace.

lkiesow committed:
MH-12502, Add Options To Get Unique Files From Workspace

This patch adds an optional `uniqueFilename` parameter to
`workspace.get()`, allowing for the requested file to be stored in a
unique location which makes it safe to modify the files content as well
as deleting the file when it is not used anymore.

lkiesow committed:
MH-12502, Convert Workspace Read To InputStream

This patch makes the `workspace.read()` method return an InputStream,
ensuring that the files content which may be read directly from the
working file repository (if possible) is not modified.

In case the files cannot be read directly (e.g. non-local files), the
files are temporarily copied into the workspace, but will automatically
be removed once the returned InputStream is closed to not leave any
files lying around in the workspace.

karendolan commented:
Hi Lars, I added a comment on the Jira ticket, I'm curious why it's better for the workspace to be cleaned of mediapackage elements immediately instead of at end of a workflow. Is it because items are more often loaded into the workspace outside of workflows?

lkiesow commented:
Please find my answer in the Jira ticket.

lkiesow commented:
Note that this ticket is currently [next in line](https://github.com/opencast/opencast/pulls?utf8=%E2%9C%93&q=is%3Apr+is%3Aopen+base%3Adevelop+sort%3Acreated-asc+) on `develop` and can thus be merged (while that rule still stands) if you have no further comments on this. If there are any open issues with this, please let me know.

smarquard commented:
Sorry for the delay. I aim to review this in the next few days.

lkiesow commented:
Cool, thanks

smarquard commented on modules/search-service-impl/src/main/java/org/opencastproject/search/impl/solr/SolrIndexManager.java:
```diff
@@ -693,14 +694,18 @@ static String mkString(Collection<?> as, String sep) {
-      File f = workspace.get(cat.getURI());
+      f = workspace.get(cat.getURI(), true);
       in = new FileInputStream(f);
       return mpeg7CatalogService.load(in);
```
Since this wants an InputStream, we could do workspace.read(cat.getURI()) here

smarquard replied:
@lkiesow, do you want to change that in this PR? There are now 2 downstream PRs that depend on this one, so would be good to get it merged.

staubesv replied:
@smarquard Please answer @lkiesow 's question above so that we can proceed.

smarquard replied:
Which question? Or did you get me and Lars the wrong way around here? I think I'm waiting for him.

staubesv replied:
@smarquard Sorry, I was too fast in reading this ;-)

@lkiesow You owe Stephen an answer here.

lkiesow replied:
Sorry for the delay. I've added a patch to use `Workspace::read` instead of creating a temporary file in the workspace.

smarquard reviewed (commented):
I deployed and tested this branch with three operations: rebuild the admin-ng index, create an event with uploaded file and fast testing workflow, and edited the metadata.

This branch reduces the number of files left behind in workspace and the WFR after these operations, and everything ran as expected.

There is one possible change to loadMpeg7Catalog which would be a slight simplification of the code, although I'd also be happy to merge this as-is if you want to leave that for a later PR.

smarquard commented:
For reference for future work, here were the files left behind after these 3 operations (removing all contents of workspace and WFR between operations). This is an all-in-one build.

After a reindex (8)

```
workspace/http_devubuopc301.uct.ac.za_8081/assets/assets/3e6a5d42-60ac-46e7-986b-aa03c070a49e/security-policy-episode/0/xacml.xml
workspace/http_devubuopc301.uct.ac.za_8081/assets/assets/3e6a5d42-60ac-46e7-986b-aa03c070a49e/security-policy-episode/1/xacml.xml
workspace/http_devubuopc301.uct.ac.za_8081/assets/assets/cb4fd64c-f671-4d51-8700-5f3d139bc5bd/security-policy-episode/11/xacml.xml
workspace/http_devubuopc301.uct.ac.za_8081/assets/assets/37186748-45e8-4bb6-aede-9a832b30bc21/security-policy-episode/0/xacml.xml
workspace/http_devubuopc301.uct.ac.za_8081/assets/assets/5449166d-45d8-4b30-878c-d222be42bfa1/security-policy-episode/0/xacml.xml
workspace/http_devubuopc301.uct.ac.za_8081/assets/assets/71adb26c-a4df-4867-ac48-7ff810e020d2/security-policy-episode/0/xacml.xml
workspace/http_devubuopc301.uct.ac.za_8081/assets/assets/71adb26c-a4df-4867-ac48-7ff810e020d2/security-policy-episode/3/xacml.xml
workspace/http_devubuopc301.uct.ac.za_8081/assets/assets/e2ddcd57-d85b-48e7-9b9e-0bda42080965/security-policy-episode/0/xacml.xml
```

After a workflow (6)

```
files/mediapackage/a8e721f3-6480-4dd3-94f6-a3c98364acbb/security-policy-episode/xacml.xml
files/mediapackage/a8e721f3-6480-4dd3-94f6-a3c98364acbb/security-policy-episode/xacml.xml.md5

workspace/http_devubuopc301.uct.ac.za_8081/assets/assets/a8e721f3-6480-4dd3-94f6-a3c98364acbb/security-policy-episode/0/xacml.xml
workspace/http_devubuopc301.uct.ac.za_8081/static/mh_default_org/engage-player/a8e721f3-6480-4dd3-94f6-a3c98364acbb/security-policy-episode/xacml.xml
workspace/http_devubuopc301.uct.ac.za_8081/static/mh_default_org/engage-player/a8e721f3-6480-4dd3-94f6-a3c98364acbb/9be8981d-b314-4e87-b1f2-18c363a6fa71/dublincore.xml
workspace/mediapackage/a8e721f3-6480-4dd3-94f6-a3c98364acbb/security-policy-episode/xacml.xml
```

After editing metadata of a published event (5)

```
files/mediapackage/a8e721f3-6480-4dd3-94f6-a3c98364acbb/9be8981d-b314-4e87-b1f2-18c363a6fa71/dublincore.xml
files/mediapackage/a8e721f3-6480-4dd3-94f6-a3c98364acbb/9be8981d-b314-4e87-b1f2-18c363a6fa71/dublincore.xml.md5
workspace/http_devubuopc301.uct.ac.za_8081/assets/assets/a8e721f3-6480-4dd3-94f6-a3c98364acbb/security-policy-episode/0/xacml.xml
workspace/http_devubuopc301.uct.ac.za_8081/assets/assets/a8e721f3-6480-4dd3-94f6-a3c98364acbb/security-policy-episode/1/xacml.xml
workspace/mediapackage/a8e721f3-6480-4dd3-94f6-a3c98364acbb/9be8981d-b314-4e87-b1f2-18c363a6fa71/dublincore.xml
```

lkiesow committed:
MH-12502, Read MPEG7 catalogs from working file repository

This patch switches to `Workspace::read` for loading MPEG7 catalogs in
the search service to avoid copying files into the workspace if
possible.

lkiesow commented:
My guess is that most of the remaining files are caused by the XACMLAuthorizationService. But that's just a guess. Not sure if I have time to look into it. At least not before the conference.

*Update*: Had a short look at it. After @wsmirnow's pull request #46 is merged, only `hasPermission(…)` will load the ACL files into the workspace. Since it seems that the JBoss library used actually needs the files in the workspace we cannot do much about that except for using `get(…, true)` for now. I will create a pull request for that once this one and pull request #46 are merged.
